### ISSUE TITLE
Presence of underbar in adapter name prevents it's use in URI

### ISSUE THREAD

**Issue by pivotal-slate** [rsim/oracle-enhanced#500]
Underbar is not permitted in URI schemes, so `oracle_ehanced` cannot be to specify the adapter name in a URI scheme.  That is, `oracle_ehanced://...` will generate an error when parsed, so it can't be specified in `DATABASE_URL`.

**Comment by akil-rails** [comment]
DATABASE_URL is also an alternate way in Rails 4 http://edgeguides.rubyonrails.org/configuring.html#connection-preference

**Comment by palfvin** [comment]
@akil-rails That's the problem. You can't specify "oracle_enhanced" in the scheme of DATABASE_URL because it can't be parsed.

**Comment by yahonda** [comment]
Would someone provide the actual syntax and errors for this.

**Comment by palfvin** [comment]
The full syntax is *adapter://username:password@host/dbname?parameters*, however, only `adapter` and `host` are required.

Parsing of a URI whose scheme (adapter) contains an underbar will result is an error during parsing, as in the following:

```
irb(main):014:0> URI.parse('oracle_enhanced://127.0.0.1/')
URI::InvalidURIError: bad URI(is not URI?): oracle_enhanced://127.0.0.1/
	from /Users/palfvin/.rbenv/versions/2.1.5/lib/ruby/2.1.0/uri/common.rb:176:in `split'
	from /Users/palfvin/.rbenv/versions/2.1.5/lib/ruby/2.1.0/uri/common.rb:211:in `parse'
	from /Users/palfvin/.rbenv/versions/2.1.5/lib/ruby/2.1.0/uri/common.rb:747:in `parse'
	from (irb):14
	from /Users/palfvin/.rbenv/versions/2.1.5/bin/irb:11:in `<main>'
irb(main):015:0> URI.parse('oracleenhanced://127.0.0.1/')
=> #<URI::Generic:0x007fa10b070130 URL:oracleenhanced://127.0.0.1/>
```

**Comment by cdinger** [comment]
This is an issue with how Rails parses the URI. To accommodate underscores, it looks like [activerecord substitutes `-` with `_`](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/connection_specification.rb#L36) when parsing the adapter name; `oracle-enhanced://blah/blah` appears to work.

**Comment by yahonda** [comment]
Great catch. This change made to support `ibm-db` adapter which also has a dash. 

https://github.com/rails/rails/commit/d72a0cbc807a14d3eec02a53317d11b9d9fa5815

**Comment by palfvin** [comment]
Indeed. Thank you @cdinger.  @yahonda, I assume you mean the `ibm_db` adapter, which also has an underscore. :-)

**Comment by yahonda** [comment]
yup. updated my previous commend. I think we can close this issue.

**Comment by palfvin** [comment]
Re: I think we can close this issue.
Agreed

**Comment by yahonda** [comment]
Closing this since there is a workaround found:
Use `oracle-enhanced` as a adapter name as follows.

```ruby
'oracle-enhanced://127.0.0.1/'
```
Reason: Using dash in adapter name may violate RFC then URI.parse does not allow it, Rails implements https://github.com/rails/rails/commit/d72a0cbc807a14d3eec02a53317d11b9d9fa5815 . 

Developers need to replace adapter name `oracle_enhanced` with `oracle-enhanced` to allow URI.parse to run it, then Rails replace `oracle-enhanced` with `oracle_enhanced`.

If this workaround does not solve this issue, please reopen it.

Thanks guys.

**Comment by robotex82** [comment]
This seems to break the loading of the adapter on oracle_enhanced 1.4.x, rails 3.2 and jruby 1.7.16.1 when trying to start the application:

LoadError: Please install the oracle-enhanced adapter: `gem install activerecord-oracle-enhanced-adapter` (no such file to load -- active_record/connection_adapters/oracle-enhanced_adapter)

Note the dash in /oracle-enhanced_adapter. Any ideas how fix this?

**Comment by yahonda** [comment]
rails/rails@d72a0cb  has implemented in Rails 4.2 then maybe you can workaround by implementing similar patch to the existing `connection_url_to_hash` method if you are mandatory to use DATABASE_URL.

https://github.com/rails/rails/blob/3-2-stable/activerecord/lib/active_record/connection_adapters/abstract/connection_specification.rb#L62

* From
```ruby
          adapter = config.scheme
```

* To
```
          adapter = config.scheme.gsub('-', '_')
```
